### PR TITLE
Make it possible to only scan part of a parquet file in a partition

### PR DESCRIPTION
**Pull request by yjshen**
# Which issue does this PR close?

Part of #944 

 # Rationale for this change

Open the possibility to only scan part of a parquet file in a task/partition. 

# What changes are included in this PR?
 
- Add `FileRange` to PartitionedFile.
- The file range is passed down to the `parquet` crate and filter row groups according to row groups' mid-point positions in the parquet file.

# Are there any user-facing changes?

No.

### CODE REVIEW

**Comment by alamb** [comment]
cc @tustvold

**Comment by tustvold** [comment]
I'm not sure I follow how this will work, parquet files have a block structure internally that is not amenable to seeking. Particularly with RLE data, it is common for a column chunk to consist of a single page. Could you maybe expand a bit on this?

**Comment by yjshen** [comment]
Hi @tustvold , the filter is based on row-group midpoint position. It was introduced recently in parquet crate with https://github.com/apache/arrow-rs/commit/2bca71e322fcab6c6d93a47ef71638a617e29f6c. The midpoint filtering is modeled after the [ParquetSplit](https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetInputSplit.java#L67-L91) and [MetadataConverter](https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java#L1241-L1292)

The parquet row groups level parallelism is used in MapReduce and Spark. In Spark [`splitFiles`](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/PartitionedFileUtil.scala#L26-L45) is used to generate task partitions based on partition size settings. And it may partition bigger parquet file parts to different partitions.

Currently, this PR is still WIP, since only physical plan changes are implemented. And we translate Spark physical plan to DataFusion physical plan to run natively in DataFusion https://github.com/blaze-init/spark-blaze-extension/blob/master/src/main/scala/org/apache/spark/sql/blaze/plan/NativeParquetScanExec.scala#L57-L63

**Comment by tustvold** [comment]
Oh I think I misunderstood, this is using byte ranges to filter row groups, not to filter within row groups?

**Comment by yjshen** [comment]
Yes, to filter row groups, based on RowGroup Metadata as well.

**Comment by alamb** [comment]
@liukun4515  there are already some configuration settings related to parallelism 

`target_partitions` (like degree of parallelism)
https://github.com/apache/arrow-datafusion/blob/4994eda81c2280fa78aea1ae0d92ce918947eebd/datafusion/src/execution/context.rs#L822-L823

`batch_size` (that controls how many batches are processed)
https://github.com/apache/arrow-datafusion/blob/4994eda81c2280fa78aea1ae0d92ce918947eebd/datafusion/src/execution/context.rs#L909-L914

I am not sure how well these two parameters are respected in all DataFusion operators, but I think the configurations settings are reasonable

**Comment by yjshen** [comment]
Since @tustvold is working on the new task scheduler in DataFusion, I keep this PR containing only physical plan changes. Leave planner or API unchanged.

The current changes are still helpful for downstream projects like Ballista or our Blaze, where query planning is done separately, no matter what we take later in DataFusion core. In Blaze, we use the Spark way of deciding InputSplits based on total dataset size and assigning parts of a big parquet file to multiple tasks.

@alamb @tustvold @liukun4515, please let me know what you think about this.

**Comment by tustvold** [comment]
Makes sense to me, regardless of what happens with scheduling, having a mechanism to cheaply subdivide the input streams directly, as opposed to streaming the output through a repartitioning operator, seems like a useful feature to have.

My expectation is scheduling will help with  over-provisioned parallelism in the plan, but will still need mechanisms to express that parallelism in the first place 👍

File: datafusion/core/src/datasource/listing/mod.rs [hunk 1]
```diff
@@ -32,15 +32,26 @@ pub use table::{ListingOptions, ListingTable, ListingTableConfig};
 pub type PartitionedFileStream =
     Pin<Box<dyn Stream<Item = Result<PartitionedFile>> + Send + Sync + 'static>>;
 
+/// File part identified by [start, end) positions.
```
Comment (alamb) [inline datafusion/core/src/datasource/listing/mod.rs:35]: Are these positions measured in `bytes` from the start of the file?

Comment (yjshen) [reply datafusion/core/src/datasource/listing/mod.rs:35]: Yes, get the row group midpoint in bytes in https://github.com/apache/arrow-rs/blob/master/parquet/src/file/serialized_reader.rs#L222-L232, and check if the midpoint is in the [start, end) range.

Comment (alamb) [reply datafusion/core/src/datasource/listing/mod.rs:35]:
```suggestion
/// Only scan a subset of Row Groups from the Parquet file whose data "midpoint"
/// lies within the [start, end) byte offsets. This option can be used to scan non-overlapping
/// sections of a Parquet file in parallel. 
```

**Review by alamb — COMMENTED** [review]
Looks like a good change to me. All that is missing is tests I think

My summary of this setting is that it would allow a user to get more parallelism in a plan by explicitly creating more partitions.

I believe that  @tustvold  is working on an alternate approach in  https://github.com/apache/arrow-datafusion/issues/2079 and elsewhere that would decouple the plan's parallelism from its declared number of partitions, which might make this setting less valuable

**Comment by yjshen** [comment]
@alamb @tustvold, could you please give another look at this? Thanks!

File: datafusion/core/src/physical_plan/file_format/parquet.rs [hunk 2]
```diff
@@ -537,6 +537,9 @@ fn read_partition(
                 file_metrics,
             ));
         }
+        if let Some(range) = &partitioned_file.range {
+            opt = opt.with_range(range.start, range.end);
```
Comment (Jimexist) [inline datafusion/core/src/physical_plan/file_format/parquet.rs:541]: can you check negative range or empty range?

**Comment by alamb** [comment]
🎉

**Pull request merged**